internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,414 @@
+package jobobject
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+// This file provides higher level constructs for the win32 job object API.
+// Most of the core creation and management functions are already present in "golang.org/x/sys/windows"
+// (CreateJobObject, AssignProcessToJobObject, etc.) as well as most of the limit information
+// structs and associated limit flags. Whatever is not present from the job object API
+// in golang.org/x/sys/windows is located in /internal/winapi.
+//
+// https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects
+
+// JobObject is a high level wrapper around a Windows job object. Holds a handle to
+// the job, a channel to receive iocp notifications about the lifecycle
+// of the job and a mutex for synchronized handle access.
+type JobObject struct {
+	handle     windows.Handle
+	mq         *messageQueue
+	handleLock sync.RWMutex
+}
+
+// JobLimits represents the resource constraints that can be applied to a job object.
+type JobLimits struct {
+	CPULimit           uint32
+	CPUWeight          uint32
+	MemoryLimitInBytes uintptr
+	MaxIOPS            int64
+	MaxBandwidth       int64
+}
+
+// Processor resource controls
+const (
+	ProcessorLimitMin  = 1
+	ProcessorLimitMax  = 10000
+	ProcessorWeightMin = 1
+	ProcessorWeightMax = 9
+)
+
+var (
+	ErrAlreadyClosed = errors.New("the handle has already been closed")
+	ErrNotRegistered = errors.New("job is not registered to receive notifications")
+)
+
+// SetResourceLimits sets resource limits on the job object (cpu, memory, storage).
+func (job *JobObject) SetResourceLimits(ctx context.Context, limits *JobLimits) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
```

Comment:
Wouldn't you want a Wlock here? What's the purpose of using a RWMutex?

Reply:
We only Lock when altering the handle itself, which is just during close as we invalidate the handle.

Reply:
did you mean to say "we only WLock...."? That would make sense though to only "write" lock when we update the handle.

Reply:
`RWMutex` just has `Lock` and `RLock`, where `Lock` locks for read/write, and `RLock` just locks for read.

Reply:
Yep, I deleted my previous comment. That's what I get for reading stack overflow :p